cpp/ql/lib/semmle/code/cpp/commons/NullTermination.qll

```diff
@@ -43,8 +46,9 @@ predicate mayAddNullTerminator(Expr e, VariableAccess va) {
   )
   or
   // Assignment to another stack variable
-  exists(Expr e0 |
-    mayAddNullTerminatorHelper(pragma[only_bind_into](e), va, pragma[only_bind_into](e0)) and
+  exists(StackVariable v0, Expr e0 |
+    mayAddNullTerminatorHelper(e, va, v0) and
+    mayAddNullTerminator(pragma[only_bind_into](e0), pragma[only_bind_into](v0.getAnAccess())) and
```

Comment:
Ah, I see why this speeds it up. Previously, we were computing an SCC consisting of `{mayAddNullTerminator, mayAddNullTerminatorHelper}`, but now we're computing `mayAddNullTerminatorHelper` first, and then evaluating an SCC _just_ containing `mayAddNullTerminator`. So this doesn't change any query results. Is that a correct description of what's going on?

Reply:
Yes, that's my understanding as well.  The tricky part is that we may be computing a bigger (less constrained)  `mayAddNullTerminatorHelper` relation in the first place, but I haven't seen that be a problem.